src/components/Proposal/Create/PluginForms/ABIService.ts

```diff
@@ -0,0 +1,129 @@
+import { AbiItem } from "web3-utils";
+import { Interface, isHexString } from "ethers/utils";
+import { SortService } from "lib/sortService";
+const Web3 = require("web3");
+import axios from "axios";
+import { isAddress } from "lib/util";
+
+export interface IAllowedAbiItem extends AbiItem {
+  name: string
+  type: "function"
+}
+
+export interface IAbiItemExtended extends IAllowedAbiItem {
+  action: string
+  methodSignature: string
+}
+
+const getSignatureHash = (signature: string): string => {
+  return Web3.utils.keccak256(signature).toString();
+};
+
+const getMethodSignature = ({ inputs, name }: AbiItem): string => {
+  const params = inputs?.map((x) => x.type).join(",");
+  return `${name}(${params})`;
+};
+
+const getMethodAction = ({ stateMutability }: AbiItem): "read" | "write" => {
+  if (!stateMutability) {
+    return "write";
+  }
+
+  return ["view", "pure"].includes(stateMutability) ? "read" : "write";
+};
+
+const getMethodSignatureAndSignatureHash = (method: AbiItem,): { methodSignature: string; signatureHash: string } => {
+  const methodSignature = getMethodSignature(method);
+  const signatureHash = getSignatureHash(methodSignature);
+  return { methodSignature, signatureHash };
+};
+
+const isAllowedABIMethod = ({ name, type }: AbiItem): boolean => {
+  return type === "function" && !!name;
+};
+
+/**
+ * Given valid ABI returns write functions with all thier data.
+ * @param {AbiItem[]} abi
+ */
+export const extractABIMethods = (abi: AbiItem[]): IAbiItemExtended[] => {
+  const allowedAbiItems = abi.filter(isAllowedABIMethod) as IAllowedAbiItem[];
+
+  return allowedAbiItems.map((method): IAbiItemExtended => ({
+    action: getMethodAction(method),
+    ...getMethodSignatureAndSignatureHash(method),
+    ...method,
+  }))
+    .filter((method) => method.action === "write")
+    .sort(({ name: a }, { name: b }) => SortService.evaluateString(a, b, 1));
+};
+
+/**
+ * Given array of ABI parameters objects, returns true if all values are valid.
+ * Data example:
+ * [{ type: address, value: "0x25112235dDA2F775c81f0AA37a2BaeA21B470f65" }]
+ * @param {array} data
+ * @returns {boolean}
+ */
+export const validateABIInputs = (data: Array<any>): boolean => {
+  for (const input of data) {
+    switch (input.type) {
+      case "address":
+        if (!isAddress(input.value)) {
+          return false;
+        }
+        break;
+      case "bytes":
+        if (!isHexString(input.value)) {
+          return false;
+        }
+        break;
+      case "uint256":
+        if (/^\d+$/.test(input.value) === false) {
+          return false;
+        }
+        break;
+    }
+  }
+  return true;
+};
+
+/**
+ * Given contract address returns it's ABI data.
+ * @param {string} contractAddress
+ */
+export const getABIByContract = async (contractAddress: string): Promise<Array<any>> => {
+  const url = `https://api.etherscan.io/api?module=contract&action=getabi&address=${contractAddress}&apikey=${process.env.ETHERSCAN_API_KEY}`;
```

Comment:
need support for different etherscan urls for different networks (kovan and rinkeby) 
e.g https://api-rinkeby.etherscan.io...for rinkeby.

for xdai , use blockscount.com api see https://blockscout.com/poa/xdai/api-docs

Reply:
Done.